lib/_http_client.js

```diff
@@ -646,8 +646,15 @@ function requestOnPrefinish() {
     responseKeepAlive(req);
 }
 
-function emitFreeNT(socket) {
-  socket.emit('free');
+function emitFreeNT(req) {
+  req.emit('close');
+  if (req.res) {
+    req.res.emit('close');
+  }
```

Comment:
Had to move the close emitters here to ensure they are emitted after all userland 'end' handlers have completed.